### PR TITLE
Dependency Convergence Fix for ANTLR

### PR DESCRIPTION
**Pull request by isaki-x**
This PR does two things:

**Enable Maven Enforcement**

I have enabled two built in rules, the first is for Java version (I've set a minimum of Java 8 as the code base does utilize lambda) and the second for dependency convergence checks.

**Exclude antlr-runtime for stringtemplate in Maven managed dependencies**

This seems to fix the issue where a circular dependency is introduced by requiring 3.4 of ANTLR while requiring 4.0.2 for string template.

Please see my notes here: #1182

### CODE REVIEW

**Comment by isaki-x** [comment]
This PR has turned into 4 changes:

- Added enforcer plugin support.
- Corrected circular dependency for antlr.
- Upgraded Dakka from 0.9.16 to 09.17.
- Explicitly configured maven JAR plugin.

File: pom.xml [hunk 1]
```diff
@@ -489,6 +495,19 @@
                         </excludes>
                     </configuration>
                 </plugin>
+                <plugin>
+                    <groupId>org.apache.maven.plugins</groupId>
+                    <artifactId>maven-jar-plugin</artifactId>
+                    <executions>
+                        <execution>
```
Comment (stevenschlansker) [inline pom.xml]: What's the purpose of configuring `default-jar` here?

Comment (isaki-x) [reply pom.xml]: It isn't really required as I don't think JDBI builds any shaded jars. The reason it is there is I lifted it from another POM I have which sets other settings for the plugin as well and this was required to get everything working for that particular build. I removed most of the custom settings and neglected to remove this one. I will remove it.

**Comment by isaki-x** [comment]
@stevenschlansker Good catch on the excess configuration for the jar plugin; I have committed a corrected version of the POM for your review.

File: pom.xml [hunk 2]
```diff
@@ -601,6 +612,28 @@
...
+                <executions>
+                    <execution>
+                        <id>enforce-jdbi</id>
+                        <goals>
+                            <goal>enforce</goal>
+                        </goals>
+                        <configuration>
+                            <rules>
+                                <!-- Ensure consistency -->
+                                <dependencyConvergence />
```
Comment (qualidafial) [inline pom.xml]: We already have a different plugin for this purpose from `basepom`. The enforcer plugin doesn't provide very helpful messages, unfortunately.

File: pom.xml [hunk 3]
```diff
@@ -392,6 +392,12 @@
                 <groupId>org.antlr</groupId>
                 <artifactId>stringtemplate</artifactId>
                 <version>4.0.2</version>
+                <exclusions>
+                    <exclusion>
+                        <artifactId>antlr-runtime</artifactId>
+                        <groupId>org.antlr</groupId>
+                    </exclusion>
+                </exclusions>
```
Comment (qualidafial) [inline pom.xml]: I'm still unclear why this exclusion would be necessary, since we're managing the `antlr-runtime` dependency version on line 225.

File: pom.xml [hunk 4]
```diff
@@ -489,6 +495,11 @@
                         </excludes>
                     </configuration>
                 </plugin>
+                <plugin>
+                    <!-- Fixes missing plugin during CI builds -->
+                    <groupId>org.apache.maven.plugins</groupId>
+                    <artifactId>maven-jar-plugin</artifactId>
+                </plugin>
```
Comment (qualidafial) [inline pom.xml]: Let's try to keep separate concerns in separate PRs. Something like this is easy to review and approve in isolation. In practice it takes much longer to review and approve PRs when multiple concerns are conflated into the same work.

File: pom.xml (hunk 3 above)
Comment (isaki-x) [reply pom.xml]: It is because of a circular dependency. You end up with two versions of the same library being pulled in. The dependency management seems to fall over on a circular dependency on a managed dependency. If you remove this exclusion block (and you are DropWizard, or you have the enforcer plugin enabled yourself, which I have done), the build fails. `dropwizard-jdbi3` does not build on anything newer than 3.2.1 of JDBI as a result of the issue this fixes.

File: pom.xml (hunk 4 above)
Comment (isaki-x) [reply pom.xml]: Sure, I can move this. This fixed the fact CI was failing which is why I included it. I can put in a separate branch for the CI fixes.

File: pom.xml (hunk 2 above)
Comment (isaki-x) [reply pom.xml]: I'm unfamiliar with `basepom`, however it clearly doesn't handle dependency convergence the same way as the enforcer plugin as it allowed a circular dependency reference to make it through (breaking downstream things, like DropWizard, that rely on dependency convergence to prevent bugs/issues). Perhaps there is a way to configure it so that it plays ball with the standard maven plugins used by projects that depend on JDBI3?

**Comment by isaki-x** [comment]
Please see #1186 for a PR with just the CI fixes. Once that is resolved, those changes will disappear from this PR. I do not wish to remove them from here as that will cause the CI to start failing again.

**Comment by stevenschlansker** [comment]
We merged in that PR, this one can now be rebased

**Comment by isaki-x** [comment]
I will reopen with a clean history. The rebase brought in too much noice.
